File: Problems/0054. (M) Spiral Matrix/Python3/solution.py
class Solution:
  def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
    # walk edges and shrink

    ret = []
    m, n = len(matrix[0]), len(matrix)
    l, u = 0, 0
    r, d = m, n
    while l < r-1 and u < d-1:
      for i in range(l, r):
        ret.append(matrix[u][i])
      for i in range(u+1, d):
        ret.append(matrix[i][r-1])
      for i in range(r-2, l-1, -1):
        ret.append(matrix[d-1][i])
      for i in range(d-2, u, -1):
        ret.append(matrix[i][l])
      l += 1
      r -= 1
      u += 1
      d -= 1
    
    if len(ret) < m*n:
      if n > m:
        for i in range(u, d):
          ret.append(matrix[i][l])
      else:
        for i in range(l, r):
          ret.append(matrix[u][i])

    return ret

File: Problems/0054. (M) Spiral Matrix/Java/solution.java
class Solution {
  public List<Integer> spiralOrder(int[][] matrix) {
    // walk edges and shrink

    List<Integer> ret = new ArrayList();
    int m = matrix[0].length;
    int n = matrix.length;
    int l = 0;
    int u = 0;
    int r = m;
    int d = n;
    while (l < r-1 && u < d-1) {
      for (int i = l; i < r; i++)
        ret.add(matrix[u][i]);
      for (int i = u+1; i < d; i++)
        ret.add(matrix[i][r-1]);
      for (int i = r-2; i > l-1; i--)
        ret.add(matrix[d-1][i]);
      for (int i = d-2; i > u; i--)
        ret.add(matrix[i][l]);
      l++;
      r--;
      u++;
      d--;
    }
    if (ret.size() < m*n) {
      if (n > m) {
        for (int i = u; i < d; i++)
          ret.add(matrix[i][l]);
      }
      else {
        for (int i = l; i < r; i++)
          ret.add(matrix[u][i]);
      }
    }
    return ret;
  }
}
